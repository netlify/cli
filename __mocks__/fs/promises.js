// eslint-disable-next-line no-undef, @typescript-eslint/no-require-imports
const { fs } = require('memfs')

// eslint-disable-next-line no-undef
module.exports = fs.promises
