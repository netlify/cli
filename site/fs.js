'use strict'

const fs = require('fs').promises
const path = require('path')
const util = require('util')

const rimraf = require('rimraf')

const rimrafAsync = util.promisify(rimraf)

const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch (_) {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = path.join(src, item)
      const destPath = path.join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(path.dirname(filePath), { recursive: true })
  } catch (_) {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}

module.exports = {
  copyDirRecursiveAsync,
  ensureFilePathAsync,
  removeRecursiveAsync,
}
