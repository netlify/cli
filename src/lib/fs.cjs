// @ts-check
const {
  constants,
  promises: { access, readFile, rm, stat },
} = require('fs')
const { version } = require('process')

const del = require('del')
const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */
const readFileAsyncCatchError = async (filePath) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */
const rmdirRecursiveAsync = async (path) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')

module.exports = {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
