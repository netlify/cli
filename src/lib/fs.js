const fs = require('fs')
const pathType = require('path-type')
const { promisify } = require('util')

const statAsync = promisify(fs.stat)
const readFileAsync = promisify(fs.readFile)
const writeFileAsync = promisify(fs.writeFile)
const accessAsync = promisify(fs.access)

const readFileAsyncCatchError = async filepath => {
  try {
    return { content: await readFileAsync(filepath) }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async filePath => {
  try {
    await accessAsync(filePath, fs.F_OK)
    return true
  } catch (_) {
    return false
  }
}

const isFileAsync = filePath => pathType.isFile(filePath)

module.exports = { statAsync, readFileAsync, readFileAsyncCatchError, writeFileAsync, fileExistsAsync, isFileAsync }
