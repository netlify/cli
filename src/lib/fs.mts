// @ts-check
const {
  
  constants,
  
  promises: { access, readFile, rm, stat },
} = require('fs')

const { version } = require('process')

import del from 'del'

const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */

const readFileAsyncCatchError = async (filePath: $TSFixMe) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}


const fileExistsAsync = async (filePath: $TSFixMe) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */

const rmdirRecursiveAsync = async (path: $TSFixMe) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */

const isType = async (filePath: $TSFixMe, type: $TSFixMe) => {
  try {
    const stats = await stat(filePath)
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    
    if ((error_ as $TSFixMe).code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */

const isFileAsync = (filePath: $TSFixMe) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */

const isDirectoryAsync = (filePath: $TSFixMe) => isType(filePath, 'isDirectory')

export default {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
