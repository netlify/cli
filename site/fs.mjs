import { promises as fs } from 'fs'
import { join, dirname } from 'path'
import { promisify } from 'util'

import rimraf from 'rimraf'

const rimrafAsync = promisify(rimraf)

export const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = join(src, item)
      const destPath = join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

export const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(dirname(filePath), { recursive: true })
  } catch {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

export const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}
