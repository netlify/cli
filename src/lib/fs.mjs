// @ts-check
import { constants } from 'fs'
import { access, stat } from 'fs/promises'

export const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
export const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
export const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')
