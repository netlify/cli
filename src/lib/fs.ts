import { constants } from 'fs'
import { access, stat } from 'fs/promises'

const isErrnoException = (value: unknown): value is NodeJS.ErrnoException =>
  value instanceof Error && Object.hasOwn(value, 'code')

export const fileExistsAsync = async (filePath: string) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 */
const isType = async (filePath: string, type: 'isFile' | 'isDirectory') => {
  try {
    const stats = await stat(filePath)
    if (type === 'isFile') return stats.isFile()
    return stats.isDirectory()
  } catch (error) {
    if (isErrnoException(error) && error.code === 'ENOENT') {
      return false
    }

    throw error
  }
}

/**
 * Checks if the provided filePath is a file
 */
export const isFileAsync = async (filePath: string) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 */
export const isDirectoryAsync = async (filePath: string) => isType(filePath, 'isDirectory')
