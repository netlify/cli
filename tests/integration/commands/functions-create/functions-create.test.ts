import { readFile } from 'fs/promises'

import execa from 'execa'
import { describe, expect, test } from 'vitest'

import { fileExistsAsync } from '../../../../src/lib/fs.mjs'
import cliPath from '../../utils/cli-path.cjs'
import { answerWithValue, CONFIRM, DOWN, handleQuestions } from '../../utils/handle-questions.cjs'
import { getCLIOptions, withMockApi } from '../../utils/mock-api.cjs'
import { withSiteBuilder } from '../../utils/site-builder.cjs'

describe.concurrent('functions:create command', () => {
  test('should create a new function directory when none is found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir', async (builder) => {
      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Enter the path, relative to your site',
          answer: answerWithValue('test/functions'),
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/test/functions/hello-world/hello-world.js`)).toBe(true)
      })
    })
  })

  test('should create a new edge function directory when none is found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir', async (builder) => {
      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: CONFIRM,
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
        {
          question: 'What route do you want your edge function to be invoked on?',
          answer: answerWithValue('/test'),
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/netlify/edge-functions/hello/hello.js`)).toBe(true)
      })
    })
  })

  test('should use specified edge function directory when found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-custom-edge-functions-dir', async (builder) => {
      builder.withNetlifyToml({ config: { build: { edge_functions: 'somethingEdgy' } } })

      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: CONFIRM,
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
        {
          question: 'What route do you want your edge function to be invoked on?',
          answer: answerWithValue('/test'),
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/somethingEdgy/hello/hello.js`)).toBe(true)
      })
    })
  })

  test('should install function template dependencies on a site-level `package.json` if one is found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir-with-package-json', async (builder) => {
      builder.withPackageJson({
        packageJson: {
          dependencies: {
            '@netlify/functions': '^0.1.0',
          },
        },
      })

      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Enter the path, relative to your site',
          answer: answerWithValue('test/functions'),
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: answerWithValue(`set-cookie`),
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        const { dependencies } = JSON.parse(await readFile(`${builder.directory}/package.json`, 'utf-8'))

        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/set-cookie.js`)).toBe(true)
        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/package.json`)).toBe(false)
        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/package-lock.json`)).toBe(false)
        expect(typeof dependencies.cookie).toBe('string')

        expect(dependencies['@netlify/functions']).toBe('^0.1.0')
      })
    })
  })

  test('should install function template dependencies in the function sub-directory if no site-level `package.json` is found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir-without-package-json', async (builder) => {
      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Enter the path, relative to your site',
          answer: answerWithValue('test/functions'),
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: answerWithValue(`set-cookie`),
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/set-cookie.js`)).toBe(true)
        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/package.json`)).toBe(true)
        expect(await fileExistsAsync(`${builder.directory}/test/functions/set-cookie/package-lock.json`)).toBe(true)
      })
    })
  })

  test('should not create a new function directory when one is found', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-functions-dir', async (builder) => {
      builder.withNetlifyToml({ config: { build: { functions: 'functions' } } })

      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Select the language of your function',
          answer: CONFIRM,
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(cliPath, ['functions:create'], getCLIOptions({ apiUrl, builder }))

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/functions/hello-world/hello-world.js`)).toBe(true)
      })
    })
  })

  test('should only show function templates for the language specified via the --language flag, if one is present', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir', async (builder) => {
      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Enter the path, relative to your site',
          answer: answerWithValue('test/functions'),
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(
          cliPath,
          ['functions:create', '--language', 'javascript'],
          getCLIOptions({ apiUrl, builder }),
        )

        handleQuestions(childProcess, createFunctionQuestions)

        await childProcess

        expect(await fileExistsAsync(`${builder.directory}/test/functions/hello-world/hello-world.js`)).toBe(true)
      })
    })
  })

  test('throws an error when the --language flag contains an unsupported value', async () => {
    const siteInfo = {
      admin_url: 'https://app.netlify.com/sites/site-name/overview',
      ssl_url: 'https://site-name.netlify.app/',
      id: 'site_id',
      name: 'site-name',
      build_settings: { repo_url: 'https://github.com/owner/repo' },
    }

    const routes = [
      {
        path: 'accounts',
        response: [{ slug: 'test-account' }],
      },
      { path: 'sites/site_id/service-instances', response: [] },
      { path: 'sites/site_id', response: siteInfo },
      {
        path: 'sites',
        response: [siteInfo],
      },
      { path: 'sites/site_id', method: 'patch', response: {} },
    ]

    await withSiteBuilder('site-with-no-functions-dir', async (builder) => {
      await builder.buildAsync()

      const createFunctionQuestions = [
        {
          question: "Select the type of function you'd like to create",
          answer: answerWithValue(DOWN),
        },
        {
          question: 'Enter the path, relative to your site',
          answer: answerWithValue('test/functions'),
        },
        {
          question: 'Pick a template',
          answer: CONFIRM,
        },
        {
          question: 'Name your function',
          answer: CONFIRM,
        },
      ]

      await withMockApi(routes, async ({ apiUrl }) => {
        const childProcess = execa(
          cliPath,
          ['functions:create', '--language', 'coffeescript'],
          getCLIOptions({ apiUrl, builder }),
        )

        handleQuestions(childProcess, createFunctionQuestions)

        await expect(childProcess).rejects.toThrowError('Invalid language: coffeescript')

        expect(await fileExistsAsync(`${builder.directory}/test/functions/hello-world/hello-world.js`)).toBe(false)
      })
    })
  })
})
