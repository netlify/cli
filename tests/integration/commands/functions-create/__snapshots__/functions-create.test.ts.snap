// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functions:create command > fixture: nx-integrated-monorepo > should create a new edge function 1`] = `
"[[edge_functions]]
function = \\"abtest\\"
path = \\"/test\\""
`;
